### PR TITLE
BOM-1614 : OEP-18 and Python 3.8

### PR DESCRIPTION
**Pull request by mzulqarnain1**
- Added Python 3.8 compatibility
- Dropped Python 2.7 support
- OEP-18 related changes

Relevant JIRA : https://openedx.atlassian.net/browse/BOM-1614

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -12,3 +12,12 @@ quality:
 
 package:
 	python setup.py register sdist upload
+
+upgrade: export CUSTOM_COMPILE_COMMAND=make upgrade
+upgrade: ## update the requirements/*.txt files with the latest packages satisfying requirements/*.in
+	# pip install -q -r requirements/pip_tools.txt
```
Comment (awais786) [inline Makefile:18]: remove this.

Comment (mzulqarnain1) [reply Makefile]: done.

**Comment by awais786** [comment]
@morenol Please review

File: requirements/base.in [hunk 2]
```diff
@@ -0,0 +1,8 @@
+# Core requirements for using this package
+-c constraints.txt
+
+edx-opaque-keys
+PyContracts
+sphinx
```
Comment (morenol) [inline requirements/base.in:6]: I think that in this file we only should have the dependencies that were in setup.py before this change, namely:
PyContracts, edx-opaque-keys and xblock, the other two seems to me that should be in a doc.in file or something like that.

File: tox.ini [hunk 3]
```diff
@@ -1,7 +1,22 @@
...
+[testenv:quality]
+deps =
+    -r requirements/test.txt
+commands =
+    pycodestyle --config=pycodestyle
+	pylint edx_user_state_client
+
+[testenv:docs]
+deps =
+    -r requirements/test.txt
```
Comment (morenol) [inline tox.ini:18]: If we create a doc.txt that should be here too.

File: requirements/base.in (hunk 2 above)
Comment (mzulqarnain1) [reply requirements/base.in]: done.

**Comment by mzulqarnain1** [comment]
> I little suggestion, after that I think that it is ready

@morenol Done. Please review again.

**Pull request merged**